gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApplicationServiceImpl.java

```diff
@@ -524,15 +529,26 @@ private static void validateAndEncodeClientCertificate(
     ) {
         if (applicationSettings.getTls() != null && !StringUtils.isBlank(applicationSettings.getTls().getClientCertificate())) {
             // validate certificate
+            final Certificate[] certificates;
             try {
-                final Certificate[] certificates = KeyStoreUtils.loadPemCertificates(applicationSettings.getTls().getClientCertificate());
-                // For some cases, KeyStoreUtils does not throw an exception but simply returns an empty array of certificates.
-                if (certificates.length == 0) {
-                    throw new InvalidApplicationCertificateException("An error has occurred while parsing client certificate");
-                }
+                certificates = KeyStoreUtils.loadPemCertificates(applicationSettings.getTls().getClientCertificate());
             } catch (Exception e) {
-                throw new InvalidApplicationCertificateException("An error has occurred while parsing client certificate");
+                throw new ApplicationInvalidCertificateException();
+            }
+            // For some cases, KeyStoreUtils does not throw an exception but simply returns an empty array of certificates.
+            if (certificates.length == 0) {
+                throw new ApplicationEmptyCertificateException();
+            }
+            if (certificates.length > 1) {
+                throw new ApplicationMultipleCertificatesException();
             }
+
+            // Accept only client certificates.
+            final boolean isCertificateAuthority = ((X509Certificate) certificates[0]).getBasicConstraints() != -1;
+            if (isCertificateAuthority) {
```

Comment:
There is also a constraint on key usage that we could add to enforce client certificate checks.
Can you look at the `certificates[0].getKeyUsage()` and see if it is appropriate to add some additional check to make sure the certificate is a client certificate and nothing else?

Reply:
Looks like the key usage aren't mandatory so no need to add extra checks